### ISSUE TITLE
[HOLD for payment 2023-02-02] [$1000] Add payment message comes only after sometime when the bank is removed.

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14262]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Go to settings -> payment
2. Make sure you have only one payment method enabled
3. Delete the payment method

## Expected Result
Add payment message should come immediately

## Actual Result
Add payment message comes after a delay

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.54-0
**Reproducible in staging?:** y
**Reproducible in production?:** y 
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:**  

https://user-images.githubusercontent.com/43996225/212206335-bb100d71-d23c-4312-b8e4-bc03433eb2c2.mp4

https://user-images.githubusercontent.com/43996225/212206345-6d108446-f74a-4c0f-8b81-5b2cac6c93e5.mov

**Expensify/Expensify Issue URL:**
**Issue reported by:** @abdulrahuman5196
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1673543185153959

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01c49a234eb88604e9</li>
        <li>Upwork Job ID: 1616099123063414784</li>
        <li>Last Price Increase: 2023-01-19</li>
    </ul>
</details>

**Comment by maddylewis** [comment]
@kavimuru - is this a duplicate report of https://github.com/Expensify/App/issues/14046#issuecomment-1374445380 ?

let me know - thanks!

**Comment by maddylewis** [comment]
it sounds like this proposed solution could work as a fix for this - https://github.com/Expensify/App/issues/14046#issuecomment-1374445380

**Comment by abdulrahuman5196** [comment]
Copy pasting the original solution from https://github.com/Expensify/App/issues/14046#issuecomment-1374445380

Root Cause
In PaymentMethodList.js we have set the add payment method message to be displayed as 'ListEmptyComponent' which will only display when we provide empty data to FlatList. But in this case when we delete the payment method and go offline, we still have the payment method data with "pendingAction": "delete" value as noted below which will only go away after we get response from server(also request still in network queue as well). So we still pass a data with payment entry to FlatList and flatList doesn't show the add payment menthod message.

{
  "title": "Plaid Checking",
  "description": "Account ending in 0000",
  "methodID": 1854845,
  "key": "bankAccount-1854845",
  "accountType": "bankAccount",
  "isDefault": false,
  **"pendingAction": "delete",**
  "iconSize": 40,
  "iconFill": null,
  "wrapperStyle": null
}

Note: leaving extra data as not required.

**Why is the single payment data entry with pendingAction not shown?**
We send a single payment method data to Flatlist which uses OfflineWithFeedback component for each row which hides its children when it has pendingAction of 'delete' - https://github.com/Expensify/App/blob/main/src/components/OfflineWithFeedback.js#L90. So visually the row doesn't have anything.


--- a/src/pages/settings/Payments/PaymentMethodList.js
+++ b/src/pages/settings/Payments/PaymentMethodList.js
@@ -4,6 +4,7 @@ import PropTypes from 'prop-types';
 import {FlatList} from 'react-native';
 import lodashGet from 'lodash/get';
 import {withOnyx} from 'react-native-onyx';
+import {withNetwork} from '../../../components/OnyxProvider';
 import styles from '../../../styles/styles';
 import * as StyleUtils from '../../../styles/StyleUtils';
 import MenuItem from '../../../components/MenuItem';
@@ -122,6 +123,10 @@ class PaymentMethodList extends Component {
             combinedPaymentMethods = _.filter(combinedPaymentMethods, paymentMethod => paymentMethod.accountType === this.props.filterType);
         }
 
+        if (!this.props.network.isOffline) {
+            combinedPaymentMethods = _.filter(combinedPaymentMethods, paymentMethod => paymentMethod.pendingAction !== CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE);
+        }
+
         combinedPaymentMethods = _.map(combinedPaymentMethods, paymentMethod => ({
             ...paymentMethod,
             onPress: e => this.props.onPress(e, paymentMethod.accountType, paymentMethod.accountData, paymentMethod.isDefault, paymentMethod.methodID),
@@ -254,6 +259,7 @@ PaymentMethodList.defaultProps = defaultProps;
 
 export default compose(
     withLocalize,
+    withNetwork(),
     withOnyx({
         bankAccountList: {
             key: ONYXKEYS.BANK_ACCOUNT_LIST,

This UI filter prevents any pending action to be sent to the flatlist which results in this issue. We can also add errors check in the filter if the reviewer thinks that would make solution fool proof. More information on the original proposal - https://github.com/Expensify/App/issues/14046#issuecomment-1373634035

**Comment by thienlnam** [comment]
Still reproducible on main, looks like we already have a couple good proposals here for review cc @eVoloshchak 


https://user-images.githubusercontent.com/30609178/213533983-a5e34c7e-9c38-4e80-a860-112f61a0cabd.mov

**Comment by thienlnam** [comment]
Actually, @jasperhuangg would you consider this something that should have been resolved in https://github.com/Expensify/App/issues/14046? Or should we have it taken care of in this issue

**Comment by abdulrahuman5196** [comment]
@thienlnam 
Just to note: I was one for proposal at https://github.com/Expensify/App/issues/14046#issuecomment-1374445380 (But my proposal was not implemented though)and reporter of this issue. The core issue behind https://github.com/Expensify/App/issues/14046 is **“Add payment message never shows up when switching from offline to online as it was not removed from onyx”** and current issue is **"Add payment message only shows up after sometime when user is online itself(after successful API call) because we don't have a UI filter"**

**Comment by thienlnam** [comment]
Yup, I'm aware of that - but this flow should have been tested while getting fixed in the previous issue and could just be considered a regression of the other issue.

I do see that you made comment of this in your other proposal, and so I agree that we can take this as a separate issue.

**Comment by thienlnam** [comment]
While both proposals are similar, I'm going to move forward with @s77rt 's proposal as it prevents another edge case where we have an error deleting the member, but still need it to show up while the user is offline so they can clear the error

**Comment by abdulrahuman5196** [comment]
@thienlnam  I also initially called out the errors check in my proposal
https://github.com/Expensify/App/issues/14262#issuecomment-1397202047

>This UI filter prevents any pending action to be sent to the flatlist which solves the issue. We can also add errors check and other pending actions check in the filter if required.

Even called out the same in original proposal on #14046

**Comment by thienlnam** [comment]
Ah my bad missed that comment, please make sure to add it in your code changes next time - sorry @s77rt

**Comment by abdulrahuman5196** [comment]
Sure thank you. @thienlnam

**Comment by abdulrahuman5196** [comment]
@thienlnam  PR is ready for the fix - https://github.com/Expensify/App/pull/14483
@eVoloshchak

**Comment by maddylewis** [comment]
@abdulrahuman5196 / @eVoloshchak - have you applied to the job in Upwork so that we can process payment? 

https://www.upwork.com/jobs/~01c49a234eb88604e9

**Comment by eVoloshchak** [comment]
Applied

**Comment by abdulrahuman5196** [comment]
Applied

**Comment by abdulrahuman5196** [comment]
Gentle reminder @maddylewis

**Comment by maddylewis** [comment]
thanks! i was OOO at the end of last week. let me process payments now 👍

**Comment by maddylewis** [comment]
@eVoloshchak @abdulrahuman5196  - offers sent!

**Comment by abdulrahuman5196** [comment]
@maddylewis Thank you. Accepted the offer.

**Comment by maddylewis** [comment]
kk, both payments have been processed 👍

**Comment by eVoloshchak** [comment]
@maddylewis, could you please check if we're eligible for the bonus payment?
I think the PR was merged within the 3 business days

**Comment by maddylewis** [comment]
yep, looks like you both are. let me send the bonuses.

**Comment by maddylewis** [comment]
paid!

**Comment by maddylewis** [comment]
buddy check for regression test here - https://expensify.slack.com/archives/C01SKUP7QR0/p1675798332736439

**Comment by maddylewis** [comment]
@eVoloshchak and @thienlnam - lmk if we're good to check off your action items in this comment - https://github.com/Expensify/App/issues/14262#issuecomment-1405494372

in the meantime, ill get a regression test added asap.

**Comment by eVoloshchak** [comment]
- The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/10333
- The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/10333#issuecomment-1429537475

- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

I the third item is the same as in https://github.com/Expensify/App/issues/13615#issuecomment-1386358515
> For this, I don't really think there was a regression or we could've caught this earlier. This behavior was introduced back when we originally added workspaces into newDot. We have solid documentation on ensuring actions are done optimistically, so I think that should be sufficient in this case.
